File: manage.mjs
#!/usr/bin/env node
import { spawnSync, spawn } from 'child_process';
import readline from 'readline';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT_DIR = __dirname;
const APP_DIR = path.join(ROOT_DIR, 'app');

const isWin = process.platform === 'win32';
const npmCmd = isWin ? 'npm.cmd' : 'npm';

function printHeader(title) {
  console.log('\n' + '='.repeat(45));
  console.log(`  ${title}`);
  console.log('='.repeat(45));
}

function runDev() {
  printHeader('🚀 啟動本機開發伺服器 (Next.js Dev Server)');
  const child = spawn(npmCmd, ['run', 'dev'], {
    cwd: APP_DIR,
    stdio: 'inherit',
    shell: true,
  });
  child.on('exit', (code) => {
    console.log(`\n🛑 開發伺服器已結束 (退出碼: ${code})`);
  });
}

function runBuild() {
  printHeader('🔨 執行建置與型別檢查 (Build Check)');
  const res = spawnSync(npmCmd, ['run', 'build'], {
    cwd: APP_DIR,
    stdio: 'inherit',
    shell: true,
  });
  if (res.status === 0) {
    console.log('\n✅ 建置與型別檢查成功！');
  } else {
    console.log(`\n❌ 建置失敗，退出碼: ${res.status}`);
  }
}

function runPush(commitMsg) {
  printHeader('📤 推送程式碼至 GitHub');
  const statusRes = spawnSync('git', ['status', '--porcelain'], {
    cwd: ROOT_DIR,
    encoding: 'utf-8',
  });

  if (!statusRes.stdout || !statusRes.stdout.trim()) {
    console.log('ℹ️ 目前沒有任何變更需要提交。');
    return;
  }

  console.log('📄 偵測到的變更清單：');
  console.log(statusRes.stdout.trim().split('\n').map(l => `   ${l}`).join('\n'));
  console.log();

  const doCommitAndPush = (msg) => {
    const finalMsg = msg || 'feat: update project';
    console.log('➕ 執行 git add . ...');
    spawnSync('git', ['add', '.'], { cwd: ROOT_DIR, stdio: 'inherit' });

    console.log(`📝 執行 git commit -m "${finalMsg}" ...`);
    spawnSync('git', ['commit', '-m', finalMsg], { cwd: ROOT_DIR, stdio: 'inherit' });

    console.log('🚀 執行 git push origin main ...');
    const pushRes = spawnSync('git', ['push', 'origin', 'main'], { cwd: ROOT_DIR, stdio: 'inherit' });

    if (pushRes.status === 0) {
      console.log('\n🎉 成功推送到 GitHub！');
    } else {
      console.log(`\n❌ 推送失敗，退出碼: ${pushRes.status}`);
    }
  };

  if (commitMsg) {
    doCommitAndPush(commitMsg);
  } else {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question('💬 請輸入 Commit 訊息 (留空預設為 "feat: update project"): ', (ans) => {
      rl.close();
      doCommitAndPush(ans.trim());
    });
  }
}

function runStatus() {
  printHeader('📊 查看 Git 狀態');
  spawnSync('git', ['status'], { cwd: ROOT_DIR, stdio: 'inherit' });
}

function showInteractiveMenu() {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  console.log('\n' + '━'.repeat(35));
  console.log('   🌟 MyKcal 專案管理工具');
  console.log('━'.repeat(35));
  console.log('  [1] 🚀 本機執行測試 (npm run dev)');
  console.log('  [2] 🔨 建置與檢查 (npm run build)');
  console.log('  [3] 📤 推送至 GitHub (git push)');
  console.log('  [4] 📊 查看狀態 (git status)');
  console.log('  [0] 🚪 退出');
  console.log('━'.repeat(35));

  rl.question('請選擇操作 [0-4]: ', (choice) => {
    rl.close();
    const c = choice.trim();
    if (c === '1') runDev();
    else if (c === '2') runBuild();
    else if (c === '3') runPush();
    else if (c === '4') runStatus();
    else if (c === '0') console.log('👋 再見！');
    else console.log('⚠️ 無效的選擇。');
  });
}

const args = process.argv.slice(2);
if (args.length > 0) {
  const cmd = args[0].toLowerCase();
  if (['dev', 'start', 'run'].includes(cmd)) runDev();
  else if (['build', 'check'].includes(cmd)) runBuild();
  else if (['push', 'deploy'].includes(cmd)) runPush(args.slice(1).join(' '));
  else if (['status', 'st'].includes(cmd)) runStatus();
  else {
    console.log(`
使用說明：
  node manage.mjs              啟動互動式選單
  node manage.mjs dev          啟動本機測試伺服器
  node manage.mjs build        執行建置檢查
  node manage.mjs push [訊息]   自動 commit 並推送到 GitHub
  node manage.mjs status       查看 Git 狀態
`);
  }
} else {
  showInteractiveMenu();
}

File: manage.py
#!/usr/bin/env python3
"""
MyKcal 專案自動化輔助腳本
支援功能：
  1. 本機執行測試 (Dev Server)
  2. 程式碼建置與 TypeScript 檢查 (Build Check)
  3. 自動 Commit & 推送至 GitHub (Git Push)
  4. 查看當前 Git 與專案狀態 (Status)
"""

import sys
import os
import subprocess
import shutil

ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
APP_DIR = os.path.join(ROOT_DIR, 'app')

def print_header(text: str):
    print("\n" + "=" * 50)
    print(f"  {text}")
    print("=" * 50)

def get_npm_command():
    if shutil.which('npm.cmd'):
        return 'npm.cmd'
    if shutil.which('npm'):
        return 'npm'
    return 'npm'

def run_dev():
    """啟動本機開發伺服器"""
    print_header("🚀 啟動本機開發伺服器 (Next.js Dev Server)")
    npm_cmd = get_npm_command()
    try:
        subprocess.run([npm_cmd, "run", "dev"], cwd=APP_DIR, check=True)
    except KeyboardInterrupt:
        print("\n\n🛑 已停止開發伺服器。")
    except Exception as e:
        print(f"\n❌ 執行失敗: {e}")

def run_build():
    """執行建置與 TypeScript 檢查"""
    print_header("🔨 執行建置與型別檢查 (Build Check)")
    npm_cmd = get_npm_command()
    try:
        result = subprocess.run([npm_cmd, "run", "build"], cwd=APP_DIR)
        if result.returncode == 0:
            print("\n✅ 建置與型別檢查成功！")
        else:
            print(f"\n❌ 建置失敗，退出碼: {result.returncode}")
        return result.returncode == 0
    except Exception as e:
        print(f"\n❌ 執行失敗: {e}")
        return False

def run_push(commit_msg: str = ""):
    """自動加入變更、Commit 並推送到 GitHub"""
    print_header("📤 推送程式碼至 GitHub")
    
    # Check git status first
    status_proc = subprocess.run(["git", "status", "--porcelain"], cwd=ROOT_DIR, capture_output=True, text=True)
    if not status_proc.stdout.strip():
        print("ℹ️ 目前沒有任何變更需要提交。")
        return

    print("📄 偵測到的變更清單：")
    for line in status_proc.stdout.strip().split('\n'):
        print(f"   {line}")
    print()

    if not commit_msg:
        commit_msg = input("💬 請輸入 Commit 訊息 (留空預設為 'feat: update project'): ").strip()
        if not commit_msg:
            commit_msg = "feat: update project"

    try:
        print("➕ 執行 git add . ...")
        subprocess.run(["git", "add", "."], cwd=ROOT_DIR, check=True)

        print(f"📝 執行 git commit -m \"{commit_msg}\" ...")
        subprocess.run(["git", "commit", "-m", commit_msg], cwd=ROOT_DIR, check=True)

        print("🚀 執行 git push origin main ...")
        push_res = subprocess.run(["git", "push", "origin", "main"], cwd=ROOT_DIR)

        if push_res.returncode == 0:
            print("\n🎉 成功推送到 GitHub！")
        else:
            print(f"\n❌ 推送失敗，退出碼: {push_res.returncode}")
    except subprocess.CalledProcessError as e:
        print(f"\n❌ Git 操作中斷: {e}")
    except Exception as e:
        print(f"\n❌ 發生錯誤: {e}")

def run_status():
    """查看 Git 狀態"""
    print_header("📊 查看 Git 狀態")
    subprocess.run(["git", "status"], cwd=ROOT_DIR)

def interactive_menu():
    while True:
        print("\n" + "━" * 40)
        print("   🌟 MyKcal 專案管理工具")
        print("━" * 40)
        print("  [1] 🚀 本機執行測試 (npm run dev)")
        print("  [2] 🔨 建置與檢查 (npm run build)")
        print("  [3] 📤 推送至 GitHub (git push)")
        print("  [4] 📊 查看狀態 (git status)")
        print("  [0] 🚪 退出")
        print("━" * 40)
        
        choice = input("請選擇操作 [0-4]: ").strip()
        if choice == '1':
            run_dev()
        elif choice == '2':
            run_build()
        elif choice == '3':
            run_push()
        elif choice == '4':
            run_status()
        elif choice == '0':
            print("👋 再見！")
            break
        else:
            print("⚠️ 無效的選擇，請重新輸入。")

def main():
    if len(sys.argv) > 1:
        cmd = sys.argv[1].lower()
        if cmd in ['dev', 'start', 'run']:
            run_dev()
        elif cmd in ['build', 'check']:
            run_build()
        elif cmd in ['push', 'deploy']:
            msg = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else ""
            run_push(msg)
        elif cmd in ['status', 'st']:
            run_status()
        elif cmd in ['help', '-h', '--help']:
            print("""
使用說明：
  python manage.py              啟動互動式選單
  python manage.py dev          啟動本機測試伺服器
  python manage.py build        執行建置檢查
  python manage.py push [訊息]   自動 commit 並推送到 GitHub
  python manage.py status       查看 Git 狀態
""")
        else:
            print(f"⚠️ 未知指令: {cmd}，輸入 python manage.py --help 查看用法。")
    else:
        interactive_menu()

if __name__ == '__main__':
    main()
